Export all public URLs to the environment

philippelatulippe opened the pull request:
To support workflows where multiple artifacts have an install page, I've had this step export two new environment variables that contain all public URLs, instead of just the last one. The URLs are accessible through:
* a pipe-separated array of URLs, `BITRISE_PUBLIC_INSTALL_PAGE_URL_LIST`
* a map of files and URLS, `BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP.` Entries are separated by pipes (|), keys and values a colon (:)

This fixes #43 and #22

philippelatulippe commented:
Using a colon (:) as the separator between key and value was a bad idea considering the value will always contain a colon. I'm not sure what would be best, an asterisk, an exclamation point, the greater than sign.

kdobmayer commented:
Hi @philippelatulippe !

Thanks for the PR. Do we really need the `BITRISE_PUBLIC_INSTALL_PAGE_URL_LIST` variable? The map seems to be enough and we can parse the url list and files from it.

The separator between keys and values could be a comma `,` or a colon followed by a space. But a simple colon is not bad, just split the value at the first colon to get the file and the url.

kdobmayer commented on main.go:
```diff
@@ -221,6 +230,29 @@ func main() {
+		if err := tools.ExportEnvironmentWithEnvman("BITRISE_PUBLIC_INSTALL_PAGE_URL_LIST", value); err != nil {
+			fail("Failed to export BITRISE_PUBLIC_INSTALL_PAGE_URL_LIST, error: %s", err)
+		}
+		log.Printf("An array of public install page urls is now available in the Environment Variable: BITRISE_PUBLIC_INSTALL_PAGE_URL_LIST (value: %s)", value)
```
Put a newline after logging the env vars to be more readable.

philippelatulippe replied:
Do you mean to print a newline to the log or to put a newline in the code?

philippelatulippe commented:
Thanks for reviewing my merge request. :) I kept `BITRISE_PUBLIC_INSTALL_PAGE_URL_LIST` to make things simple for people who just want an unordered list of URLs and don't care which is which. It's particularly nice for people using shell scripts, which has poor string processing facilities. I wouldn't mind removing it though, it would simplify the code.

That's also why I agonized over the key/value separator. I'll switch it to a comma, it seems ideal.

viktorbenei commented:
I agree @philippelatulippe re the `_LIST`, that it would be great for Bash scripts where string manipulation might be a problem, but even then it shouldn't be just a list of the URLs IMO as in that case there's absolutely no indication which URL is for which artifact. E.g. sending the list to Slack, no one would know which URL is what.

The map solution seems to fit this better, at least for now. Won't be "that nice" on slack but it's manageable I think.

Or, alternatively the `_LIST` one could also be something like: `filename1:URL1|filename2:URL2`, so basically a simplified version of the map.

viktorbenei commented:
My bad, I thought the `_MAP` is in JSON, but checked the code again and it's exactly what I described :D

So in that case I'd +1 to only have the `_MAP`, as the `_LIST` with only the URLs seem to have limited utility.

philippelatulippe commented:
I've fixed the issues from the review:
* Don't export an array of files, only the map
* print a newline after logging the value

I've also used a comma (,) as the key/value separator, since it makes it easier to use a split function to get the key and value.

philippelatulippe commented:
I accidentally pushed an old version of my code, I've now amended the commit.

kdobmayer commented:
You can iterate over the public page urls with the following code snippet:

```shell
# example apk paths and urls
BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP="/path/file.apk,https://url.com|/path/file2.apk,https://url2.com"

while read -r line; do
  echo "public page url: ${line#*,}"
done <<< "$(echo "${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP}" | tr '|' '\n')"
```

You can use any command instead of `echo`.
Or put all of them into an array and use it later:

```shell
BITRISE_PUBLIC_INSTALL_PAGE_URL_LIST=()
while read -r line; do
  BITRISE_PUBLIC_INSTALL_PAGE_URL_LIST+=("${line#*,}")
done <<< "$(echo "${BITRISE_PUBLIC_INSTALL_PAGE_URL_MAP}" | tr '|' '\n')"
```

It's also working with colon (`:`) as a separator between paths and urls.

kdobmayer commented:
Hi @philippelatulippe ,

We discussed it and decided to use only the map as output. I will merge your PR then remove the list output.

philippelatulippe commented:
I thought I had already removed the map output. Did I forget to push? Or
perhaps I forgot to update the yaml. I'm not currently at my laptop so I
can't fix it right now, I'd be glad if you merged and fixed it, but
otherwise I can fix it later this afternoon.

viktorbenei commented:
Merged & released, thanks for the PR @philippelatulippe ! ;)
